### ISSUE TITLE
Remove the RuntimeWarning in pcoa

### ISSUE THREAD

**Issue by johnchase** [biocore/scikit-bio#1410]
A warning like the following appears any time `skbio.stats.ordination.pcoa` is run:
```
RuntimeWarning: The result contains negative eigenvalues. Please compare their magnitude with
the magnitude of some of the largest positive eigenvalues. If the negative ones are smaller, it's 
probably safe to ignore them, but if they are large in magnitude, the results won't be useful. See 
the Notes section for more details. The smallest eigenvalue is -0.004237399024826933 and the 
largest is 2.359730572801951.
  RuntimeWarning```

Is it possible to add a flag to suppress the warning?

**Comment by gregcaporaso** [comment]
@rob-knight just checked with a few statisticians on this. I agree that this would be really good to get rid of. I can try to summarize that conversation (I was cc'ed), unless @rob-knight can get to it first.

**Comment by gregcaporaso** [comment]
cc @Jorge-C

**Comment by jairideout** [comment]
As a workaround, it is easy to suppress warnings in a very targeted manner:

https://docs.python.org/3.5/library/warnings.html#temporarily-suppressing-warnings

**Comment by Jorge-C** [comment]
That warning shows up iff the input distances can't be perfectly embedded
in the euclidean/output space. If the magnitude of the smallest negative
eigenvalue (i.e., abs(min(eigvals))) is big relative to the largest
positive one, the embedding in euclidean space becomes inaccurate, and you
shouldn't draw conclusions from that embedding/it doesn't represent the
actual distances. When the difference is 3 orders of magnitude, the
embedding should be pretty good. I don't think Legendre&Legendre gave any
"boundary" for what ratio makes the embedding good enough, so that's why it
just prints it to let the user make a decision. If there's a way to compute
the loss of accuracy we could stop printing it if the loss is less than x%.

2016-08-10 12:09 GMT-06:00 Greg Caporaso <notifications@github.com>:

**Comment by gregcaporaso** [comment]
@rob-knight got the following input from [Shyamal Peddada](http://www.niehs.nih.gov/research/atniehs/labs/bb/staff/peddada/). @Jorge-C, do you have thoughts on this?  

> The standard PCA are based on the sample covariance matrix.  Mathematically, the sample covariance is positive definite unless the dimension is larger than the sample size.  If the dimension is larger than the sample size then the smallest eigenvalues will be zero due to singularity.  However, due to numerical reasons (round off errors), sometimes it is possible for the smallest eigenvalues to be very small negative numbers.

> However, in PCoA you are computing eigenvalues of a distance matrix.  Since the diagonal elements of a distance matrix are zero, it is necessary for at least one of its eigenvalues to be negative and at least one to be positive, i.e. the matrix is an indefinite matrix. Hence it
is a mathematical fact that you will see some negative eigenvalues - these are NOT due to numerical round off errors.

**Comment by Jorge-C** [comment]
>The standard PCA are based on the sample covariance matrix.  Mathematically, the sample covariance is positive definite unless the dimension is very large and the sample size is small.  In such a case the smallest eigenvalues will be zero due to singularity.  In such cases the numerical zeros would look like negative numbers.  But they will be very very small in absolute value.

Right. A similar thing happens with PCoA, when the distance is euclidean (or close enough): all eigenvalues are theoretically non-negative and some zero. Some of those zeros end up being numerically close to zero but negative. In our code, we deal with them and consider them to be zero, so they do not trigger the warning.


> However, in PCoA you are computing eigenvalues of a distance matrix

The eigenvalues are computed of a matrix that is not a distance matrix. The distance matrix is squared, scaled (to get what Legendre & Legendre calls the E matrix), then each element is shifted by a different amount (to get the F matrix). The eigenvalues come from this last matrix, which has non-zero diagonal.

> which is mathematically an indefinite matrix because the diagonals are zero.

The initial distance matrix is. The properties of the modified matrix depend on the distance matrix metric. For example it's positive semi-definite in the simple euclidean case (because of the relationship with PCA). It's maybe indefinite in general for other metrics like unifrac.

>  In such a case it is NOT surprising to see negative eigenvalues.  In fact, I expect you to see negative eigenvalues rather frequently when computing eigenvalues of a distance matrix.

Yes, for a distance matrix the eigenvalues add up to zero, so some need to be negative if you have a positive one. The transformation to the E and F matrices breaks that limitation.

> It is a mathematical fact that you will get negative eigenvalues.  Depending upon your application, there are a variety of solutions available.  One is to obtain the "nearest" positive semidefinite matrix to the distance matrix.  It is an optimization problem. There is a closed form solution to problem, which is to replace all the negative eigenvalues by zero or with a small positive constant (to make it positive definite).  Proof is given in Rao (1984).

L&L cite a few solutions to the negative eigenvalue problem in PCoA (\S 9.2.3 of the 1998 edition, there's a newer edition too). I had a copy while I ported/wrote the code, but can't find it right now. From what I remember, one correction was named Caillez, and it looked very similar to the usual "add a scaled identity matrix to make another positive definite". There was a second option though.

There was another book (Numerical Ecology with R) that I checked while implementing that essentially says the same thing as our warning. Quoting:
"if the association coefficient used is non-Euclidean, then PCoA may react by producing several negative eigenvalues in addition to the positive ones (and a null eigenvalue in-between). The axes corresponding to negative eigenvalues cannot be represented on real ordination axes since they are complex. In most applications, this does not affect the representation of the objects on the several first principal axes, but it can lead to problems if the largest negative eigenvalues are of the same magnitude in absolute value as the first positive ones."

I didn't add a reference to it in the code/docstring because L&L looked better as a basic reference.

A possible solution is to implement the correction methods, and give the option to use them or not. The warning is still important, but maybe it could be automatically suppressed if the ratio is big enough (eg at least 10^2?).

The underlying problem is that the input distance matrix can't be embedded exactly in a euclidean space (almost never for microbial datasets). You can either tweak the distances or matrices involved to remove the negative eigenvalues (thus representing exactly a modified version of the distance matrix), or accept a slightly wrong representation of the unmodified distances. The key is how to measure that "slightly wrong", and then remove the warning if small enough (to stop worrying users).

Hope that helps! Let me know if something doesn't look right :)

**Comment by gregcaporaso** [comment]
> The warning is still important, but maybe it could be automatically suppressed if the ratio is big enough (eg at least 10^2?).

I'd be all for this - I think that would get around a lot of RuntimeWarnings that we're seeing, and seems conservative based on the *Numerical Ecology with R* quote (which I think we'd want to reference in the notes if we made this change). 

Does anyone object to this?

**Comment by ElDeveloper** [comment]
This sounds like a good idea.

**Comment by mortonjt** [comment]
I think that is reasonable.  Most of the time the error is very tiny

**Comment by antgonza** [comment]
Agree but I think it will be a good idea to add a full explanation and a reference to when this happens in the warning. Perhaps, point to a section in the documentation, this issue or both.

**Comment by Jorge-C** [comment]
I feel that the current citation is better. The section we are currently referencing (it appears in the Notes part of the docstring, which is referenced by the warning) is like half a dozen pages talking about negative eigenvalues, vs just one paragraph in the other book I cited.

**Comment by ebolyen** [comment]
I think we should parametrize the cutoff with that default, if only to make it obvious that such a cuttoff exists for people who wouldn't read the docs thoroughly.

**Comment by Jorge-C** [comment]
Yep, agree about the cutoff.
